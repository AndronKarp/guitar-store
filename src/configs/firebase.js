import firebase from "firebase";

const config = {
  apiKey: "AIzaSyDrF2CjklO0SlyV6Kv7pg7p67a4sLV3qbw",
  authDomain: "project-a0c8e.firebaseapp.com",
  databaseURL: "https://project-a0c8e.firebaseio.com",
  projectId: "project-a0c8e",
  storageBucket: "project-a0c8e.appspot.com",
  messagingSenderId: "686927330962",
  appId: "1:686927330962:web:6892625786538e6588304b",
  measurementId: "G-P6Q64LKV8Q"
};

firebase.initializeApp(config);

const database = firebase.database();
const guitarsRef = database.ref("guitars");
const auth = firebase.auth();
const usersRef = database.ref("users");
auth.getCurrentUser = () => {
  return new Promise((resolve, reject) => {
    const unsubscribe = firebase.auth().onAuthStateChanged(user => {
      unsubscribe();
      resolve(user);
    }, reject);
  });
};

export { guitarsRef, auth, usersRef };
