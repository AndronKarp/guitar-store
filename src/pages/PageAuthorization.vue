<template>
  <b-form @submit.prevent="authorize">
    <b-form-group>
      <b-form-input
        placeholder="Enter your e-mail..."
        type="email"
        v-model="$v.form.email.$model"
      ></b-form-input>
    </b-form-group>
    <b-form-group>
      <b-form-input
        placeholder="Enter your password..."
        type="password"
        v-model="$v.form.password.$model"
      ></b-form-input>
    </b-form-group>
    <b-button
      :class="{ 'bg-info': !$v.$invalid }"
      :disabled="$v.$invalid"
      type="submit"
      >Sign In</b-button
    >
  </b-form>
</template>

<script>
import { required } from "vuelidate/lib/validators";
import formValidation from "../mixins/form-validation";
import { validationMixin } from "vuelidate";
import { auth } from "../configs/firebase";

export default {
  data() {
    return {
      form: {
        email: "",
        password: ""
      }
    };
  },
  validations: {
    form: {
      email: { required },
      password: { required }
    }
  },
  methods: {
    authorize() {
      auth().signInWithEmailAndPassword(this.form.email, this.form.password);
    }
  },
  mixins: [validationMixin, formValidation]
};
</script>

<style></style>
