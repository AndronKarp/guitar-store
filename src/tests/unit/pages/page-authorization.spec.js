import { mount, createLocalVue } from "@vue/test-utils";
import PageAuthorization from "@/pages/PageAuthorization";
import BootstrapVue from "bootstrap-vue";
import { auth } from "@/configs/firebase";

jest.mock("@/configs/firebase", () => {
  return {
    auth: jest.fn().mockReturnValue({
      signInWithEmailAndPassword: jest
        .fn()
        .mockResolvedValueOnce()
        .mockResolvedValueOnce()
        .mockRejectedValueOnce()
    })
  };
});

const localVue = createLocalVue();

localVue.use(BootstrapVue);

describe("PageAuthorization.vue", () => {
  let wrapper;
  let $router;

  beforeEach(() => {
    $router = { push: jest.fn() };
    wrapper = mount(PageAuthorization, { localVue, mocks: { $router } });
  });

  test("renders authorization form", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("form submit button", () => {
    let submitButton;
    let inputs;

    beforeEach(() => {
      submitButton = wrapper.find("button");
      inputs = wrapper.findAll("input");
    });

    test("disabled if input data is not provided", () => {
      expect(submitButton.element.getAttribute("disabled")).toBe("disabled");
    });

    test("enabled if input data is provided", () => {
      inputs.setValue("value");
      wrapper.vm.$nextTick(() => {
        expect(submitButton.element.getAttribute("disabled")).toBe(null);
      });
    });
  });

  describe("methods", () => {
    describe("authorize", () => {
      test("authorizes user", async () => {
        await wrapper.vm.authorize();
        expect(auth().signInWithEmailAndPassword).toHaveBeenCalledWith(
          wrapper.vm.form.email,
          wrapper.vm.form.password
        );
      });

      test("invokes moveToHomePage method after successful authorization", async () => {
        const moveToHomePage = jest.spyOn(wrapper.vm, "moveToHomePage");
        await wrapper.vm.authorize();
        expect(moveToHomePage).toHaveBeenCalled();
      });

      test("sets authState value to false if authorization is failed", async () => {
        await wrapper.vm.authorize();
        wrapper.vm.$nextTick(() => {
          expect(wrapper.vm.authState).toBe(false);
        });
      });
    });
  });
});
