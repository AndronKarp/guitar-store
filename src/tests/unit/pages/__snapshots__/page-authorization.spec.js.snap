// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageAuthorization.vue renders authorization form 1`] = `
<div class="d-flex flex-fill justify-content-center align-items-center">
  <form class="d-flex flex-column w-25" style="min-width: 280px;">
    <fieldset class="form-group" id="__BVID__10">
      <!---->
      <div tabindex="-1" role="group" class="bv-no-focus-ring"><input type="email" placeholder="Enter your e-mail..." class="form-control" id="__BVID__11">
        <!---->
        <!---->
        <!---->
      </div>
    </fieldset>
    <fieldset class="form-group" id="__BVID__12">
      <!---->
      <div tabindex="-1" role="group" class="bv-no-focus-ring"><input type="password" placeholder="Enter your password..." class="form-control" id="__BVID__13">
        <div class="invalid-feedback">
          Invalid e-mail or password
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </fieldset> <button type="submit" disabled="disabled" class="btn btn-secondary disabled">Sign In</button>
  </form>
</div>
`;
